integration/jnifuse/fs/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java

```diff
@@ -71,50 +183,75 @@ public static NativeLibraryLoader getInstance() {
    *               on exit.
    * @throws IOException if a filesystem operation fails
    */
-  public synchronized void loadLibrary(final String tmpDir) throws IOException {
-    try {
-      System.loadLibrary(SHARED_LIBRARY_NAME);
-      LOG.info("Loaded {} by System.loadLibrary.", SHARED_LIBRARY_NAME);
-    } catch (final UnsatisfiedLinkError ule1) {
-      try {
-        System.loadLibrary(JNI_LIBRARY_NAME);
-        LOG.info("Loaded {} by System.loadLibrary.", JNI_LIBRARY_NAME);
-      } catch (final UnsatisfiedLinkError ule2) {
-        loadLibraryFromJar(tmpDir);
+  public synchronized void loadLibrary(final VersionPreference preference, final String tmpDir) throws IOException {
+
+    Optional<UnsatisfiedLinkError> err;
+
+    if (preference == VersionPreference.VERSION_2) {
+      err = load2(tmpDir);
+      if (err.isPresent()) {
+        throw err.get();
+      }
+      return;
+    }
+    if (preference == VersionPreference.VERSION_3) {
+      err = load3(tmpDir);
+      if (err.isPresent()) {
+        throw err.get();
       }
+      return;
+    }
+
+    // load libfuse2 first
+    err = load2(tmpDir);
+    if (!err.isPresent()) {
+      return;
     }
+
+    // if libfuse2 failed, load libfuse3
+    err = load3(tmpDir);
+    if (!err.isPresent()) {
+      return;
+    }
+
+    // throws if neither is loaded
+    // TODO better messaging
+    throw new UnsatisfiedLinkError();
```

Comment:
add the message?

Reply:
added a simple error message. but there is still room for improvements, like providing the details of previous loading